packages/lesswrong/server/authentication/auth0.ts

```diff
@@ -17,12 +25,42 @@ type Auth0Settings = {
 
 export const AUTH0_SCOPE = "profile email openid offline_access";
 
+type Auth0User = User<AppMetadata, UserMetadata>;
+
+abstract class IAuth0Client {
+  abstract getUserById(userId: string): Promise<Auth0User>;
+  abstract updateUserById(userId: string, data: UpdateUserData): Promise<Auth0User>;
+  abstract signupUser(email: string, password: string): Promise<void>;
+  abstract loginUser(email: string, password: string): Promise<string | null>;
+}
+
+/**
+ * We replace the real `Auth0Client` (defined below) with this mock client for
+ * end-to-end Cypress tests. This bypasses auth0 completely. Note that the mock
+ * client will always login the user, whether or not their password is correct.
+ */
+class MockAuth0Client extends IAuth0Client {
+  getUserById(_userId: string): Promise<Auth0User> {
+    throw new Error("getUserById not implemented for tests");
+  }
+
+  updateUserById(_userId: string, _data: UpdateUserData): Promise<Auth0User> {
+    throw new Error("updateUserById not implemented for tests");
+  }
+
+  async signupUser(_email: string, _password: string): Promise<void> {}
+
+  async loginUser(email: string, _password: string): Promise<string | null> {
+    return `access-token-${email}`;
+  }
+}
+
 /**
  * Auth0 Management API Client
  * For this to work, we must authorize the forum application in Auth0:
  * Applications -> APIs -> Auth0 Management API -> Machine to Machine Applications
  */
-const auth0Client = new class Auth0Client {
+class Auth0Client extends IAuth0Client {
```

Comment:
There is another `Auth0Client` in `packages/lesswrong/components/hooks/useAuth0Client.ts`, should that also use this interface?

Reply:
Oh, yeah - I forgot about that one. This interface is only for the server, and that's a frontend client, so no. I'll rename this accordingly.